### PR TITLE
[Lang] Fix numpy and taichi operations problem

### PR DESCRIPTION
**Pull request by neozhaoliang**
Related issue = #5463

Currently, we support:

1. In python scope operations between `np.ndarray/np.scalar` and `ti.Matrix` will return `ti.Maitrx`.
2. In Taichi scope operations between `np.scalar` and `ti.Matrix` will return `ti.Matrix`.
3. In Taichi scope if the first operand is `np.ndarray` and the second operand is `ti.Matrix`, the result is `ti.Matrix`. (The reverse is not supported)

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -403,6 +403,16 @@ class Matrix(TaichiOperations):
         0
     """
     _is_taichi_class = True
+    __arrar_priority__ = 1000
+
+    def __array_ufunc__(self, ufunc, method, *args, **kwargs):
```
Comment (lin-hitonami) [inline python/taichi/lang/matrix.py:408]: I'm not very familiar with numpy, so could you explain what this function does?

File: python/taichi/lang/matrix.py [hunk 2]
```diff
@@ -403,6 +403,16 @@ class Matrix(TaichiOperations):
         0
     """
     _is_taichi_class = True
+    __arrar_priority__ = 1000
```
Comment (lin-hitonami) [inline python/taichi/lang/matrix.py:406]:
```suggestion
    __array_priority__ = 1000
```

File: python/taichi/lang/matrix.py (hunk 1 above)
Comment (neozhaoliang) [reply python/taichi/lang/matrix.py:408]: It's here: https://numpy.org/neps/nep-0018-array-function-protocol.html

If the user's custom array implements this interface, NumPy will call this function to handle the operations between ndarray and user's custom arrays.

File: tests/python/test_numpy.py [hunk 3]
```diff
@@ -213,3 +213,22 @@ def func2(a: ti.types.ndarray()):
...
+
+
+@test_utils.test()
+def test_numpy_op_with_matrix():
+    a = np.cos(1)
+    b = ti.Vector([1, 2])
+    c = np.array([0, 1])
+    x = a + b
+    y = c + b
+    assert isinstance(x, ti.lang.matrix.Matrix) and isinstance(
```
Comment (ailzhang) [inline tests/python/test_numpy.py:225]:
```suggestion
    assert isinstance(x, ti.Matrix) and isinstance(
```

File: tests/python/test_numpy.py [hunk 4]
```diff
@@ -213,3 +213,22 @@ def func2(a: ti.types.ndarray()):
...
+        y, ti.lang.matrix.Matrix)
```
Comment (ailzhang) [inline tests/python/test_numpy.py:226]:
```suggestion
        y, ti.Matrix)
```

File: tests/python/test_numpy.py [hunk 5]
```diff
@@ -213,3 +213,22 @@ def func2(a: ti.types.ndarray()):
...
+
+    @ti.kernel
+    def test():
+        x = a + b
```
Comment (ailzhang) [inline tests/python/test_numpy.py:230]: can we also check on the values in `x`?

File: python/taichi/lang/matrix.py [hunk 6]
```diff
@@ -403,6 +403,16 @@ class Matrix(TaichiOperations):
...
     _is_taichi_class = True
+    __array_priority__ = 1000
+
+    def __array_ufunc__(self, ufunc, method, *args, **kwargs):
+        """Make sure operations between a numpy ndarray and Taichi matrix always return a Matrix.
+        This function is called only when the first operand is a numpy ndarray.
+        """
+        if method == "__call__":
+            name = ufunc.__name__
+            return getattr(self, f"__{name}__")(args[0].tolist(), **kwargs)
```
Comment (ailzhang) [inline python/taichi/lang/matrix.py:414]: would you mind adding a test with 2d numpy array with ti.Matrix?

**Review by ailzhang — COMMENTED** [review]
Thanks!

File: python/taichi/lang/matrix.py (hunk 6 above)
Comment (neozhaoliang) [reply python/taichi/lang/matrix.py:414]: 2d test added

File: tests/python/test_numpy.py [hunk 7]
```diff
@@ -215,11 +215,26 @@ def func2(a: ti.types.ndarray()):
...
+    arr2d = np.array([[0, 1], [2, 3]])
+    mat = ti.Matrix([[0, 1], [2, 3]])
 
-    x = np.array([1, 2, 3])
-    assert foo(x, 1) == 2
+    assert isinstance(scalar + vec, ti.Matrix) and isinstance(
+        mat + arr2d, ti.Matrix)
+
+    @ti.kernel
+    def test():
```
Comment (ailzhang) [inline tests/python/test_numpy.py:230]: https://docs.taichi-lang.org/docs/reference#the-assert-statement 
IIRC assert only works when debug=True?

Comment (neozhaoliang) [reply tests/python/test_numpy.py:230]: Thanks for this detailed review! I have added `debug=True` in the test.

**Comment by neozhaoliang** [comment]
removed the support for operations between `numpy.ndarray` and `ti.Matrix` since the implementation is too hacky and does not work for binary ops like `div` and `mod`.

**Review by lin-hitonami — APPROVED** [review]
LGTM!

**Comment by neozhaoliang** [comment]
Thanks!

**Pull request merged**